stdlib/string.gr

```diff
@@ -1297,3 +1297,111 @@ export let decode = (bytes: Bytes, encoding: Encoding) => {
+ */
+@disableGC
+export let forEachCodePoint = (f: (Number) -> Void, s: String) => {
+  let (>>>) = WasmI32.shrU
+  let (-) = WasmI32.sub
+  let (&) = WasmI32.and
+  let (<) = WasmI32.ltU
+  let (<=) = WasmI32.leU
+  let (==) = WasmI32.eq
+  let (+) = WasmI32.add
+
+  let s = WasmI32.fromGrain(s)
+
+  let byteSize = WasmI32.load(s, 4n)
+
+  let mut ptr = s + 8n
+  let end = ptr + byteSize
+  
+  let mut idx = 0n
+  while (ptr < end) {
+    let byte = WasmI32.load8U(ptr, 0n)
+    let codePointByteCount = if ((byte & 0x80n) == 0x00n) {
+      1n
+    } else if ((byte & 0xF0n) == 0xF0n) {
+      4n
+    } else if ((byte & 0xE0n) == 0xE0n) {
+      3n
+    } else {
+      2n
+    }
+
+    // Note that even if up to 4 bytes are needed to represent Unicode
+    // codepoints, this doesn't mean 32 bits. The highest allowed code point is
+    // 0x10FFFF and it should not change in future versions of Unicode. Which
+    // means no more than 21 bits are necessary to represent a code point and
+    // thus we can use Grain's "simple" numbers that hold up to 31 bits and
+    // avoid heap allocations. I'm assuming here that getCodePoint would throw
+    // MalformedUnicode exception for values exceeding this limit.
+    let codePoint = getCodePoint(ptr)
+    f(tagSimpleNumber(codePoint))
+
+    ptr += codePointByteCount
+    idx += 1n
+  }
+}
+
+/**
+ * Iterates over Unicode code points in this string. This same as the
+ * forEachCodePoint function, but with a code point index provided for
+ * convenience.
+ * 
+ * @param f: (Number,Number) -> Void - The iterator function
+ * @param s: String - The string to iterate
+ * @returns Void
+ */
+@disableGC
+export let forEachCodePointi = (f: (Number,Number) -> Void, s: String) => {
+  let (>>>) = WasmI32.shrU
+  let (-) = WasmI32.sub
+  let (&) = WasmI32.and
+  let (<) = WasmI32.ltU
+  let (<=) = WasmI32.leU
+  let (==) = WasmI32.eq
+  let (+) = WasmI32.add
+
+  let s = WasmI32.fromGrain(s)
+
+  let byteSize = WasmI32.load(s, 4n)
+
+  let mut ptr = s + 8n
+  let end = ptr + byteSize
+  
+  let mut idx = 0n
+  while (ptr < end) {
+    let byte = WasmI32.load8U(ptr, 0n)
+    let codePointByteCount = if ((byte & 0x80n) == 0x00n) {
+      1n
+    } else if ((byte & 0xF0n) == 0xF0n) {
+      4n
+    } else if ((byte & 0xE0n) == 0xE0n) {
+      3n
+    } else {
+      2n
+    }
+
+    // Note that even if up to 4 bytes are needed to represent Unicode
+    // codepoints, this doesn't mean 32 bits. The highest allowed code point is
+    // 0x10FFFF and it should not change in future versions of Unicode. Which
+    // means no more than 21 bits are necessary to represent a code point and
+    // thus we can use Grain's "simple" numbers that hold up to 31 bits and
+    // avoid heap allocations. I'm assuming here that getCodePoint would throw
+    // MalformedUnicode exception for values exceeding this limit. Idem for the
+    // index. String can never have more than 2^31 codepoints (or bytes for
+    // that matter).
```

Comment:
```suggestion
    // MalformedUnicode exception for values exceeding this limit.
```

Technically a string's length is an unsigned int, so Grain theoretically can support strings up to 2^32 in length. We'll probably add a hard upper limit though.